test/unit/org/apache/cassandra/cql3/CQLTester.java

```diff
@@ -970,7 +976,8 @@ protected static void schemaChange(String query)
 
             QueryOptions options = QueryOptions.forInternalCalls(Collections.<ByteBuffer>emptyList());
 
-            lastSchemaChangeResult = statement.executeLocally(queryState, options);
+            ResultMessage result = statement.executeLocally(queryState, options);
+            recordedSchemaChangeResult = recordSchemaChange ? result : recordedSchemaChangeResult;
```

Comment:
Using a ternary operator here is a bit confusing. An `if` would be clearer.

Reply:
agree, but mostly to avoid updating with the same value; can we use

```
if (recordSchemaChange)
    recordedSchemaChangeResult = result;
```

it is more clear, but also avoids a race condition issue if a user updates the schema from multiple threads; we should avoid setting as we could set an old history by mistake.

Reply:
@dcapwell Your comment made me pause. I always assumed that this code will be called in a single threaded way (outside of the tearDown logic). If a user updates the schema from multiple threads there are no way to avoid some flakiness with `assertLastSchemaChange` because we cannot know which thread will update the `recordedSchemaChangeResult` last.
If we want the code to be more robust, I think we should consider removing the volatile variable and let the user store the result of the schema change on his side.
We are exploring that approach with @bereng. If it lead to something we will post a patch next week.

Reply:
> If we want the code to be more robust, I think we should consider removing the volatile variable and let the user store the result of the schema change on his side.

The common case with CQLTester is a heavy stateful set of things with utilities to fetch the last modified X (table, alter, etc.), given this I do find the added method is enough for the common case, as `protected static void schemaChange(String query, boolean recordSchemaChange)` allows you to opt out when this behavior isn't desired.

Now, a new method which lets you have more control is +1 from me, but don't feel we should refactor existing tests to use it (though 100% ok if schemaChange delegates to this new method), the thread safety of the tests is only single threaded tests, so these methods tend to work well for that case.

Can you explain more what this new logic is like?  Is it any different than

```
// method now returns ResultMessage
protected static ResultMessage schemaChange(String query, boolean recordSchemaChange)
    {
        try
        {
            ClientState state = ClientState.forInternalCalls(SchemaConstants.SYSTEM_KEYSPACE_NAME);
            QueryState queryState = new QueryState(state);

            CQLStatement statement = QueryProcessor.parseStatement(query, queryState.getClientState());
            statement.validate(state);

            QueryOptions options = QueryOptions.forInternalCalls(Collections.<ByteBuffer>emptyList());

            ResultMessage result = statement.executeLocally(queryState, options);
            if (recordSchemaChange)
                 recordedSchemaChangeResult = result;
            return result;
        }
        catch (Exception e)
        {
            logger.info("Error performing schema change", e);
            throw new RuntimeException("Error setting schema for test (query was: " + query + ")", e);
        }
    }
```

Reply:
That 'new logic' @blerer is referring to is that code indeed in the new PR iiuc.